SQL CLI documentation (#1362)
astronomer/docs: astro/cli/sql-cli.md

---

:::caution 

This feature is in alpha and is not yet recommended for production workflows.

:::

Use the Astro CLI's SQL commands to run SQL workflows without having to write Python or set up Airflow. The Astro CLI handles all of the required backend configuration for running SQL locally using Airflow's reliability and observability. Using an Airflow connection, you can output the results of your SQL queries to the database of your choice. 

## Enable Astro SQL commands 

To enable Astro SQL commands, run the following command: 

```sh
astro config set -g beta.sql_cli true
```

## Create a new SQL project 

A SQL project contains the set of files necessary to run SQL queries from the Astro CLI, including YAML files for configuring your SQL environment and directories for storing SQL workflows. A SQL project can exist as its own top-level directory or within an existing Astro project. 

To create a SQL project, run the following command in either an empty directory or an existing Astro project:

```sh
astro flow init
```

The command generates the following files: 

```sh

> new_proj/

Author: @tatiana Where are the `.airflow` files here? Are they hidden?

Reviewer: Yes, @jwitz , they are hidden in the root of the project directory, in a folder `.airflow`. When the project is initilised, the folder contents are:
```
│   ├── .airflow
│   │   ├── dags
│   │   │   └── sql
│   │   ├── default
│   │   │   ├── airflow.cfg
│   │   │   └── airflow.db

```
Unless the users customised `--airflow-dags-folder` or `--airflow-home` when running `flow init`, we'll be placing the dags generated/run by the `flow` command inside `<project-dir>/.airflow/dags` and the Airflow configuration will live in the environment-specific folder.

- `workflows`: This directory contains subdirectories for SQL **workflows**. Each workflow subdirectory contains one or more related SQL files that should be run together. The default SQL project includes a few example workflows that you can run out of the box. 
- `data`: This directory contains the data for which to run your SQL workflows, formatted as `.db` files. The default SQL project includes example data for films and a simple retail platform.

## Develop your SQL project 

If you have existing `.sql` files and datasets, developing a SQL project that runs with Airflow is as simple as adding files to a few directories.

### Create SQL workflows 

To run SQL as workflows, add your `.sql` files to one of your workflow subdirectories in `workflows`. All SQL files in a given workflow run together when you run `astro flow run`. See [Run a SQL workflow](#run-a-sql-workflow).

If your `.sql` file has a `SELECT` statement, your SQL workflow will create a table in your target database named after the `.sql` file.

SQL files within a workflow can be dependent on each other using Jinja templating.

### Add data to test

> Add all data to run your SQL workflows with in the `data` directory of your project.

Author: ```suggestion
Add all data to run with your SQL workflows as `.db` files in the `data` directory of your project.
```

Author: @tatiana How does a SQL file know which `.db` file to work with? Is it defined in SQL?

Reviewer: Through the database connection properties, @jwitz !

As an example, we have this content for the `default` environment:
```
$ cat config/default/configuration.yml 
connections:
- conn_id: sqlite_conn
  conn_type: sqlite
  host: /tmp/ash/data/imdb.db
  login: null
  password: null
  schema: null
```
And this is the content for one of the workflow's files:

```
$ cat workflows/example_basic_transform/top_animations.sql 
---
conn_id: sqlite_conn
---
SELECT Title, Rating
FROM movies
WHERE Genre1=='Animation'
ORDER BY Rating desc
LIMIT 5;
```

We tell the SQL file which connection to use. If it uses SQLite locally, the connection will contain a reference to the file. If it refers to Postgres or other databases, it will contain all the properties required for those databases.

These are some examples of how to define other connections:

```## Example of Bigquery connection
#  - conn_id: bigquery_conn
#    conn_type: bigquery
#    description: null
#    extra:
#      project: "my-gcp-project"
#    host: null
#    login: null
#    password: null
#    port: null
#    schema: null

## Example of Postgres connection
#  - conn_id: postgres_conn
#    conn_type: postgres
#    host: localhost
#    schema:
#    login: postgres
#    password: postgres
#    port: 5432
#    extra:

## Example of Redshift connection
## To use this, either set the corresponding environment variables or
## replace the values started by $ inline
#  - conn_id: redshift_conn
#    conn_type: redshift
#    schema: $REDSHIFT_DATABASE
#    host: $REDSHIFT_HOST
#    port: 5439
#    login: $REDSHIFT_USERNAME
#    password: $REDSHIFT_PASSWORD

## Example of Snowflake connection
## To use this, either set the corresponding environment variables or
## replace the values started by $ inline
#  - conn_id: snowflake_conn
#    conn_type: snowflake
#    host: https://gp00000.us-east-1.snowflakecomputing.com
#    port: 443
#    login: $SNOWFLAKE_ACCOUNT_NAME
#    password: $SNOWFLAKE_PASSWORD
#    schema: $SNOWFLAKE_SCHEMA
#    extra:
#      account: "gp00000"
#      region: "us-east-1"
#      role: $SNOWFLAKE_ROLE
#      warehouse: $SNOWFLAKE_WAREHOUSE
#      database: $SNOWFLAKE_DATABASE
```

### Configure environments and Airflow connections

An environment is a set of configurations that define the Airflow environment in which your SQL will run. Each environment is defined as a subdirectory in the `config` directory containing a single `configuration.yml` file. 

An environment requires an Airflow connection to the external databases where you will load the results of your SQL workflows. You can configure one or multiple connections withing a single `configuration.yml` file. To configure a connection for an environment, add the connection's key value pairs in YAML format to the environment's `configuration.yml` file. 

See [Managing connections](https://airflow.apache.org/docs/apache-airflow/stable/howto/connection.html) for a list of connection types and required values for each type. For example, a Snowflake connection in `dev/configuration.yml` might look like the following: 

```yaml
snowflake_default:
  login: mylogin
  password: mypassword
  schema: myschema
  account: myaccount
  database: MYDATABASE
  region: us-east-1
  warehouse: MYWAREHOUSE
  role: ADMIN
```

> Note that Airflow connections configured in this directory are not accessible to your Astro project. Similarly, connections configured in your Astro project are not accessible to your SQL project. 

Author: I'm assuming this is the case @tatiana ?

Reviewer: This assumption is correct, Jake! At least for now - we'll be reviewing this in 0.3

---
sidebar_label: 'Run SQL models'
title: 'Run SQL models with the Astro CLI'
id: sql-cli
description: The Astro CLI can start up a local environment for testing SQL queries. Use YAML configuration and CLI commands to reliably run SQL in Airflow without having to write python. 
---

:::caution 

This feature is in alpha and is not yet recommended for production workflows.

:::

> Use the Astro CLI's SQL commands to run SQL workflows without having to write Python or set up Airflow. The Astro CLI handles all of the required backend configuration for running SQL locally using Airflow's reliability and observability. Using an Airflow connection, you can output the results of your SQL queries to the database of your choice. 

Reviewer: @jwitz the focus, at least for 0.1, is for us not to use Airflow observability. We are adding an abstraction layer so SQL Analysts don't feel overwhelmed by Airflow and its concepts.

That's why we intentionally removed words and concepts related to Airflow from the commands & helps. They can configure Airflow things if they know, but they don't worry about it. Users define how to connect to a database using a YML file. The fact we internally generate an Airflow connection should not be meaningful to SQL Analysts - at least not for 0.1.

## Prerequisites 

To use the Astro CLI's SQL commands, you need

> - The [Astro CLI](overview.md)

Reviewer: It is probably worth saying users need to be using version 1.7 or highter

- [Docker Desktop](https://www.docker.com/)

> - An external database service which uses one of the following databases:

Reviewer: ```suggestion
- One of the following databases:
```
Postgres and SQLite could be potentially run locally

  
    - Postgres
    - SQlite
    - BigQuery
    - Snowflake
    - Redshift

## Enable Astro SQL commands 

> To enable Astro SQL commands, run the following command: 

Reviewer: Should we mention to run this after installing the Astro CLI?

```sh
astro config set -g beta.sql_cli true
```

## Create a new SQL project 

A SQL project contains the set of files necessary to run SQL queries from the Astro CLI, including YAML files for configuring your SQL environment and directories for storing SQL workflows. A SQL project can exist as its own top-level directory or within an existing Astro project. 

To create a SQL project, run the following command in either an empty directory or an existing Astro project:

```sh
astro flow init

> ```

Reviewer: @jwitz, it is probably worth adding a disclaimer that the first time the user, this will take a bit longer than usual because it uses Docker under the hood.

We also recently noticed the following: Debian/Ubuntu users have two options:
* Use the upstream docker package, and have the possibility of having SQL projects in the `/tmp` directory
* Use the docker snap version, and do not have projects in `/tmp`

Would it make sense to have a disclaimer session here?

Author: I added a disclaimer to the prerequisites about this, I assume that we want people using the snap version primarily?

The command generates the following files: 

```sh
new_proj/
├── config
│   ├── default
│   │   └── configuration.yml
│   └── dev
│       └── configuration.yml
├── data
│   ├── imdb.db
│   └── retail.db
└── workflows
    ├── example_basic_transform
    │   └── top_animations.sql
    └── example_templating
        ├── filtered_orders.sql
        └── joint_orders_customers.sql
```

Each SQL project is composed of three directories:

- `config`: This directory contains configurations for different **environments** for running SQL. These configurations are mainly used for storing Airflow connections to external databases. Each SQL environment has one subdirectory containing exactly one file named `configuration.yml`. The default SQL project includes `default` and `dev` environments.

> - `workflows`: This directory contains subdirectories for SQL **workflows**. Each workflow subdirectory contains one or more related SQL files that should be run together. The default SQL project includes a few example workflows that you can run out of the box. 

Reviewer: ```suggestion
- `workflows`: Each subdirectory is used to define an independent SQL **workflow**. A workflow should contain one or more related SQL files that should be run together. The default SQL project includes a few example workflows you can run out of the box. 
```

> - `workflows`: This directory contains subdirectories for SQL **workflows**. Each workflow subdirectory contains one or more related SQL files that should be run together. The default SQL project includes a few example workflows that you can run out of the box. 

Reviewer: As it stands now, we expect the SQL files to contain transformations which would be outputted to a table).

> - `data`: This directory contains the data for which to run your SQL workflows, formatted as `.db` files. The default SQL project includes example data for films and a simple retail platform.

Reviewer: ```suggestion
- `data`: This directory can contain datasets to run your SQL workflows. The default SQL project includes sample SQLite databases (`.db` files) for films and a simple retail platform.
```

## Develop your SQL project 

If you have existing `.sql` files and datasets, developing a SQL project that runs with Airflow is as simple as adding files to a few directories.

### Create SQL workflows 

To run SQL as workflows, add your `.sql` files to one of your workflow subdirectories in `workflows`. All SQL files in a given workflow run together when you run `astro flow run`. See [Run a SQL workflow](#run-a-sql-workflow).

If your `.sql` file has a `SELECT` statement, your SQL workflow will create a table in your target database named after the `.sql` file.

SQL files within a workflow can be dependent on each other using Jinja templating.

### Add data to test

> Add all data to run your SQL workflows with in the `data` directory of your project.

Reviewer: This is unnecessary - it is just used in the examples so we don't have an external dependency on a database that users would have overhead to install. If they are using BQ, Snowflake, Redshift or Postgres, they would not have to add anything to the `data/` dir.

### Configure environments and Airflow connections

> An environment is a set of configurations that define the Airflow environment in which your SQL will run. Each environment is defined as a subdirectory in the `config` directory containing a single `configuration.yml` file. 

Reviewer: ```suggestion
An environment is a set of configurations that define the database environment in which your SQL will run. Each environment is a subdirectory in the `config` directory containing a single `configuration.yml` file. 
```
The environment does not necessarily care about Airflow itself. It is a way of allowing users to run workflows (locally) pointing to databases which are in different environments. It's common for analysts to develop queries using dev databases and run them in more representative databases (e.g. staging) to check the behaviour and analyse results - all this from their computer.

An environment requires an Airflow connection to the external databases where you will load the results of your SQL workflows. You can configure one or multiple connections withing a single `configuration.yml` file. To configure a connection for an environment, add the connection's key value pairs in YAML format to the environment's `configuration.yml` file. 

> See [Managing connections](https://airflow.apache.org/docs/apache-airflow/stable/howto/connection.html) for a list of connection types and required values for each type. For example, a Snowflake connection in `dev/configuration.yml` might look like the following: 

Reviewer: If possible, I would probably replace these sessions, removing any references to Airflow.

As mentioned before, I do not think we should say external databases - since the user could be running databases locally if they wanted to.

```yaml
snowflake_default:
  login: mylogin
  password: mypassword
  schema: myschema
  account: myaccount
  database: MYDATABASE
  region: us-east-1
  warehouse: MYWAREHOUSE
  role: ADMIN
```

> Note that Airflow connections configured in this directory are not accessible to your Astro project. Similarly, connections configured in your Astro project are not accessible to your SQL project. 

Reviewer: That's not quite right, Jake! At the moment, we abstract Airflow from users and create the connections for them, using another format. SQL users should not have to worry about connections or managing them in Airflow. They should only think in terms of "database connections", - which we may even prefer to rename to `database credentials.

All they should worry is about defining them in the `config/<environment>/configuration.yml` file either explicitly, eg:
```
  - conn_id: postgres_conn
    conn_type: postgres
    host: localhost
    schema:
    login: postgres
    password: postgres
    port: 5432
    extra:
```

Or implicitly, by referencing environment variables which would contain the secrets:

```
  - conn_id: redshift_conn
    conn_type: redshift
    schema: $REDSHIFT_DATABASE
    host: $REDSHIFT_HOST
    port: 5439
    login: $REDSHIFT_USERNAME
    password: $REDSHIFT_PASSWORD

```

> #### Test Airflow connections

Reviewer: ```suggestion
#### Test database connections
```

> To test all configured connections within an environment, run the following command from your SQL project directory:

Reviewer: ```suggestion
To test all configured database connections within an environment, run the following command from your SQL project directory:
```

```sh
astro flow validate --environment=<env-directory-name>
```

> This command runs Airflow's [connection testing tool](https://airflow.apache.org/docs/apache-airflow/stable/howto/connection.html#testing-connections) for all connections configured in the `configuration.yml` file of your environment subdirectory. 

Reviewer: This is an implementation detail; I don't think we should expose it to end users. What I believe is meaningful to analysts is that we validate if the user-configured credentials can establish connections to the target databases. It's a good practice to confirm this before trying to run any queries on the target database.

You can also test individual connections within an environment. For example, to test a `snowflake_default` connection in the `dev/configuration.yml`, you would run the following command to test the connection:

```sh
astro flow validate --connection=snowflake_default --environment=dev
```

## Run a SQL workflow

To run a SQL workflow in your project, run the following command: 

```sh

> astro flow run <workflow-directory> --connection=<your-connection> --environment=<your-environment>

Reviewer: @jwitz, how important is it for us to give all the flags as the first example?
The simplest way of running a workflow is simply by typing:
`astro flow run <workflow-directory>`

In this case, it defaults to the **default** environment.

> astro flow run <workflow-directory> --connection=<your-connection> --environment=<your-environment>

Reviewer: ```suggestion
astro flow run <workflow-directory> --environment=<your-environment>
```

We don't have a flag `--connection` as part of the run command (at least not in 0.1):

```
$ flow run --help
                                                                                                                               
 Usage: flow run [OPTIONS] WORKFLOW_NAME                                                                                       
                                                                                                                               
 Run a workflow locally. This task assumes that there is a local airflow DB (can be a SQLite file), that has been initialized  
 with Airflow tables.                                                                                                          
                                                                                                                               
╭─ Arguments ─────────────────────────────────────────────────────────────────────────────────────────────────────────────────╮
│ *    workflow_name      TEXT  name of the workflow directory within workflows directory. [required]                         │
╰─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────╯
╭─ Options ───────────────────────────────────────────────────────────────────────────────────────────────────────────────────╮
│ --env                              environment  environment to run in [default: default]                                    │
│ --project-dir                      PATH         (Optional) Default: current directory.                                      │
│ --verbose          --no-verbose                 Whether to show airflow logs [default: no-verbose]                          │
│ --help         -h                               Show this message and exit.                                                 │
╰─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────╯

```

The connection is set in the header of the SQL files - and the environment flag will capture from the relevant configuration file which connection to use.

```

This command automatically builds and runs a DAG based solely on the configurations in your project directory. For example, consider the following command using the default SQL project assets:

```sh

> astro flow run example_templating --connection=snowflake_default --environment=dev

Reviewer: ```suggestion
astro flow run example_templating --environment=dev
```

```

After running this command, the CLI:

- Creates a DAG based on the provided information. Because the SQL files in `example_templating` are dependent on each other, the CLI creates a DAG with two tasks based on jinja templating. The DAG is available to view in `<your-sql-project/.airflow/dags`.

> - Runs the DAG. If the SQL project exists in an Astro project, the CLI uses the Astro project's `airflow.cfg` and `airflow.db` files to start an Airflow environment. If the SQL project exists on its own, the CLI uses default `airflow.cfg` and `airflow.db` files from the SQL project. Specifically, the DAG:

Reviewer: I believe these two items are implementation details, and we should not be exposing them in the user guide - or at least not for users who do not care about them. Perhaps we could have this in more advanced configurations for existing Airflow users.

    - Connects to Snowflake using the `snowflake_default` connection configured in `dev/configuration.yml`.
    - Creates a table named `filtered_orders` and runs `filtered_orders.sql` to populate the table with data from `data/retail.db`. A table is named after the `.sql` file which includes the `SELECT` statement that creates the table. 
    - Creates a table named `joint_orders_customers` and runs `joint_orders_customers.sql` to populate the table using data from `filtered_orders`.

## Export a SQL workflow as a DAG

Export a SQL workflow as a DAG using the following command: 

```
astro flow generate <your-sql-workflow> --env=<your-env> --airflow-dags-folder=<your-dags-folder>
```

This command automatically builds a DAG that runs your workflow and exports it to `<your-dags-folder>` in your SQL project. If the DAGs directory doesn't already exist, the CLI creates it for you. 

For example, to export a DAG that runs the `example_templating` workflow based on the `dev` environment to a directory named `sql_dags`, you would run: 

```sh

> astro flow generate example_templating --env=dev --airflow-dags-folder=sql_dags

Reviewer: ```suggestion
astro flow generate example_templating --env=dev 
```
There is no flag `-airflow-dags-folder=sql_dags` as part of `generate`. If the users don't want the Airflow DAGs to be hidden, they should set the target DAGs folder directory during the project initialisation, by running:

```
flow init -airflow-dags-folder=sql_dags
```

---
sidebar_label: 'Run SQL models'
title: 'Run SQL models with the Astro CLI'
id: sql-cli

> description: The Astro CLI can start up a local environment for testing SQL queries. Use YAML configuration and CLI commands to reliably run SQL without having to write Python. 

Reviewer 2: ```suggestion
description: Use the Astro CLI to start a local environment for SQL query testing. YAML configurations and Astro CLI commands let you run SQL without writing Python code. 
```

---

:::caution 

> This feature is in alpha and is not yet recommended for production workflows.

Reviewer 2: ```suggestion
This feature is not recommended for production workflows.
```

Reviewer 3: The SQL CLI feature is in alpha, but the rest of Astro CLI 1.7 is not, AFAIK. So, would it be possible to omit references to Astro CLI 1.7 as being in alpha, in favor of just calling the SQL capabilties alpha?

Reviewer: @mikeshwe I had a chat with @sunkickr yesterday, and we don't have a stable release of 1.7 due to other features still being tested in QA / and customer feedback being collected.

:::

> Use the Astro CLI's SQL commands to run SQL workflows without having to write Python or set up Airflow. The Astro CLI handles all of the required backend configuration for running SQL locally and reliably. Using a connection configured in YAML, you can output the results of your SQL queries to the database of your choice. 

Reviewer 2: ```suggestion
Use Astro CLI SQL commands to run SQL workflows without writing Python code or setting up Airflow. The Astro CLI takes care of all of the backend configuration required for running SQL locally. A connection configured in YAML lets you output the results of your SQL queries to a specific database. 
```

## Prerequisites 

> To use the Astro CLI's SQL commands, you need

Reviewer 2: ```suggestion
To use the Astro CLI's SQL commands, you need
```

Remove this introductory statement. It's unnecessary.

> - The [Astro CLI](overview.md) version 1.7+.

Reviewer 2: ```suggestion
- The [Astro CLI](overview.md) version 1.7 or later.
```

- One of the following databases, hosted either locally or on an external service:
  
    - Postgres
    - SQlite
    - BigQuery
    - Snowflake
    - Redshift

## Enable Astro SQL commands 

To enable Astro SQL commands, run the following command: 

```sh
astro config set -g beta.sql_cli true
```

## Create a new SQL project 

A SQL project contains the set of files necessary to run SQL queries from the Astro CLI, including YAML files for configuring your SQL environment and directories for storing SQL workflows. A SQL project can exist as its own top-level directory or within an existing Astro project. 

To create a SQL project, run the following command in either an empty directory or an existing Astro project:

```sh
astro flow init
```

The command generates the following files: 

```sh
new_proj/
├── config
│   ├── default
│   │   └── configuration.yml
│   └── dev
│       └── configuration.yml
├── data
│   ├── imdb.db
│   └── retail.db
└── workflows
    ├── example_basic_transform
    │   └── top_animations.sql
    └── example_templating
        ├── filtered_orders.sql
        └── joint_orders_customers.sql
```

Each SQL project is composed of three directories:

> - `config`: This directory contains configurations for different **environments** for running SQL. These configurations are mainly used for storing connections to external databases. Each SQL environment has one subdirectory containing exactly one file named `configuration.yml`. The default SQL project includes `default` and `dev` environments.

Reviewer 2: ```suggestion
- `config`: This directory contains configurations for different environments for running SQL. These configurations  store connections to external databases. Each SQL environment has one subdirectory containing one file named `configuration.yml`. The default SQL project includes `default` and `dev` environments.
```

> - `workflows`: Each subdirectory is used to define an independent SQL **workflow**. A workflow is a subdirectory that should contain one or more related SQL files that should be run together. The default SQL project includes a few example workflows you can run out of the box. 

Reviewer 2: ```suggestion
- `workflows`: Each subdirectory is used to define an independent SQL workflow. A workflow is a subdirectory that contains one or more related SQL files that should be run together. The default SQL project includes a few example workflows you can run immediately. 
```

- `data`: This directory can contain datasets to run your SQL workflows. The default SQL project includes sample SQLite databases (`.db` files) for a film ranking and a simple retail platform.

## Develop your SQL project 

> If you have existing `.sql` files and datasets, developing a running SQL workflow is as simple as adding files to a few directories.

Reviewer 2: ```suggestion
If you have existing `.sql` files and datasets, you can develop a running SQL workflow by adding files to a few directories.
```

### Configure environments and connections

An environment is a set of configurations that define the environment in which your SQL will run. Each environment is defined as a subdirectory in the `config` directory containing a single `configuration.yml` file. 

> An environment requires a connection to the external databases where you will extract and load the results of your SQL workflows. You can configure one or multiple connections withing a single `configuration.yml` file.

Reviewer: ```suggestion
An environment requires a connection to the external databases where you will query and save the results of your SQL workflows. You can configure one or multiple connections within a single `configuration.yml` file.
```

[`dev/configuration.yml`](https://github.com/astronomer/astro-sdk/blob/main/sql-cli/include/base/config/dev/configuration.yml) contains templates for for each supported connection type. Use these templates to configure all required key-value pairs for your connection. 

For example, a Snowflake connection in `dev/configuration.yml` might look like the following: 

```yaml
snowflake_default:
  login: mylogin
  password: mypassword
  schema: myschema
  account: myaccount
  database: MYDATABASE
  region: us-east-1
  warehouse: MYWAREHOUSE
  role: ADMIN
```

> Note that connections configured in this directory are not accessible to your Astro project. Similarly, connections configured in your Astro project are not accessible to your SQL project. 

Reviewer 2: ```suggestion
Your Astro project can't access the connections configured in this directory. Similarly, your SQL project can't access connections configured in your Astro project. 
```

#### Test database connections

To test all configured databases within an environment, run the following command from your SQL project directory:

```sh
astro flow validate --environment=<env-directory-name>
```

This command runs a connection test for all databases configured in the `configuration.yml` file of your environment subdirectory. 

> You can also test individual databases within an environment. For example, to test a `snowflake_default` connection in the `dev/configuration.yml`, you would run the following command to test the connection:

Reviewer 2: ```suggestion
You can also test individual databases within an environment. For example, to test a `snowflake_default` connection in the `dev/configuration.yml`, you run the following command to test the connection:
```

```sh
astro flow validate --connection=snowflake_default --environment=dev
```

### Create SQL workflows 

To run SQL as a workflow, add your `.sql` files to one of your workflow subdirectories in `workflows`. All SQL files in a given workflow run together when you run `astro flow run`. See [Run a SQL workflow](#run-a-sql-workflow).

Each `.sql` file must have a database connection defined as `conn_id` in the front matter of the file. This front matter tells the Astro CLI where to extract and load data from. For example, a SQL query that runs against a Snowflake database would look like the following: 

```sql
---
conn_id: snowflake_default
---
SELECT Title, Rating
FROM movies
WHERE Genre1=='Animation'
ORDER BY Rating desc
LIMIT 5;
```

> If your `.sql` file has a `SELECT` statement, your SQL workflow will create a table in your target database named after the `.sql` file.

Reviewer 2: ```suggestion
If your `.sql` file includes a `SELECT` statement, your SQL workflow creates a table in your target database with the name of the `.sql` file.
```

SQL files within a workflow can be dependent on each other using Jinja templating.

:::tip 

If you define a database connection in a `.sql` file with a downstream dependency, the downstream `.sql` file will automatically inherit the database connection information. 

::: 

## Run a SQL workflow

To run a SQL workflow in your project, run the following command: 

```sh
astro flow run <workflow-directory> --environment=<your-environment>
```

This command automatically builds and runs your project as an Airflow DAG based on the configurations in your project directory. For example, consider the following command using the default SQL project assets:

```sh
astro flow run example_templating --environment=dev
```

After running this command, the CLI:

- Connects to the `data` directory using the `sqlite_conn` connection that's defined in `dev/configuration.yml` and configured in `filtered_orders.sql`.
- Creates a table named `filtered_orders` and runs `filtered_orders.sql` to populate the table with data from `data/retail.db`.
- Creates a table named `joint_orders_customers` and runs `joint_orders_customers.sql` to populate the table using data from `filtered_orders`.

> To view the implementation of this workflow in more detail, you can view the generated code using the setup steps in the following topic. 

Reviewer 2: ```suggestion
To view the implementation of this workflow in more detail, you can view the generated code. See <link>.
```

## Export a SQL workflow as a DAG

> Export a SQL workflow as an Airflow DAG using the following command: 

Reviewer 2: ```suggestion
Run the following command to export a SQL workflow as an Airflow DAG: 
```

---
sidebar_label: 'Run SQL models'
title: 'Run SQL models with the Astro CLI'
id: sql-cli
description: The Astro CLI can start up a local environment for testing SQL queries. Use YAML configuration and CLI commands to reliably run SQL without having to write Python. 
---

:::caution 

This feature is in alpha and is not yet recommended for production workflows.

:::

Use Astro CLI SQL commands to run SQL workflows without writing Python code or setting up Airflow. The Astro CLI takes care of all of the backend configuration required for running SQL locally. Configure your connections in YAML to run your SQL in a specific data warehouse.

## Prerequisites 

- The alpha version  of the [Astro CLI](overview.md) version 1.7 or later.
- [Docker Desktop](https://www.docker.com/).
- One of the following databases, hosted either locally or on an external service:
  
    - Postgres
    - SQlite
    - BigQuery
    - Snowflake
    - Redshift

:::info 

> If you use Debian or Ubuntu as your operating system, download Docker on the [Snap Store](https://snapcraft.io/docker). Otherwise, you have to [install Docker from a package](https://docs.docker.com/engine/install/ubuntu/#install-from-a-package) and will only be able to store SQL projects in your `/tmp` directory.

Reviewer: ```suggestion
If you use Debian or Ubuntu as your operating system, we recommend to [install Docker from the upstream package](https://docs.docker.com/engine/install/ubuntu/#install-from-a-package). If you install Docker using [Snap Store](https://snapcraft.io/docker), you will not be able to create projects on the `/tmp` directory.
```

:::

### Install the alpha version of Astro CLI version 1.7

Run the following command to install the alpha version of Astro CLI version 1.7:

> 

Author: @tatiana do we still want to release the docs before 1.7 is out?

> 

Reviewer: Yes, @jwitz, I spoke to @mikeshwe about this a couple of hours ago, and he also agrees!

```sh
sudo bash < <(curl -sSL https://install.astronomer.io) -s v1.7.0
```

:::warning 

The alpha version of Astro CLI version 1.7 should not be used for any development or production workflows on Astro or Astronomer Software. This version of the CLI includes updates to commands that require changes in Astronomer products that have not yet been shipped. 

To roll back to a supported version of the CLI, run the following command to uninstall the alpha:

```sh
sudo rm /usr/local/bin/astro
```

> Then, see [Install the Astro CLI](cli/install-cli.md) for steps to install the latest supported version.

Reviewer: ```suggestion
Then, see [Install the Astro CLI](cli/install-cli.md) for steps to install the latest stable version.
```

    ├── example_basic_transform
    │   └── top_animations.sql
    └── example_templating
        ├── filtered_orders.sql
        └── joint_orders_customers.sql
```

Each SQL project is composed of three directories:

- `config`: This directory contains configurations for different environments for running SQL. These configurations  store connections to external databases. Each SQL environment has one subdirectory containing one file named `configuration.yml`. The default SQL project includes `default` and `dev` environments.
- `workflows`: Each subdirectory is used to define an independent SQL workflow. A workflow is a subdirectory that contains one or more related SQL files that should be run together. The default SQL project includes a few example workflows you can run immediately.
- `data`: This directory can contain datasets to run your SQL workflows. The default SQL project includes sample SQLite databases (`.db` files) for a film ranking and a simple retail platform.

## Develop your SQL project 

If you have existing `.sql` files and datasets, you can develop a running SQL workflow by adding files to a few directories.

### Configure environments and connections

> An environment requires a connection to the external databases where you will extract and load the results of your SQL workflows. You can configure one or multiple connections withing a single `configuration.yml` file.

Reviewer 3: ```suggestion
An environment requires a connection to the external databases in which you will run your SQL workflows. You can configure one or multiple connections withing a single `configuration.yml` file.
```

[`dev/configuration.yml`](https://github.com/astronomer/astro-sdk/blob/main/sql-cli/include/base/config/dev/configuration.yml) contains templates for for each supported connection type. Use these templates to configure all required key-value pairs for your connection. 

For example, a Snowflake connection in `dev/configuration.yml` might look like the following: 

```yaml
snowflake_default:
  login: mylogin
  password: mypassword
  schema: myschema
  account: myaccount
  database: MYDATABASE
  region: us-east-1
  warehouse: MYWAREHOUSE
  role: ADMIN
```

> Your Astro project can't access the connections configured in this directory. Similarly, your SQL project can't access connections configured in your Astro project.

Reviewer 3: ```suggestion
Your Astro project currently can't access the connections configured in this directory. Similarly, your SQL project can't access connections configured in your Astro project. We plan to unify these two sets of connections in upcoming releases.
```

> Your Astro project currently can't access the connections configured in this directory. Similarly, your SQL project can't access connections configured in your Astro project. We plan to unify these two sets of connections in upcoming releases.

Author: ```suggestion
Your Astro project currently can't access the connections configured in this directory. Similarly, your SQL project can't access connections configured in your Astro project. Features for unifying these two sets of connections will be available in upcoming releases.
```
